Attention Neural Bag-of-Feature Learner for Heart Anomaly Detection (#72)
opendr-eu/opendr: docs/reference/attention_neural_bag_of_feature_learner.md

  The AF dataset was divided into 5 folds.   
- **sample_length**: *int, default=30*    
  The length of each sample (in seconds). 
  This value must be at least 30 seconds.  
  Note that this value is different from `series_length`, which is equal to `sample_length` multiplied by the sampling rate (3000 Hz).    
- **standardize**: *bool, default=True*    
  Whether to standardize the input series.   

**Returns**:  

- **train_set**: *engine.datasets.DatasetIterator*   
  The training set that is constructed from `engine.datasets.DatasetIterator` that is ready to be used with the `fit()` method.    
- **val_set**: *engine.datasets.DatasetIterator*   
  The validation set that is constructed from `engine.datasets.DatasetIterator` that is ready to be used with the `fit()` method.    
- **series_length**: *int*   
  The length of each series (the number of elements in each series).  
  This parameter is used in the constructor of `AttentionNeuralBagOfFeatureLearner`.    
- **class_weight**: *list*   
  The weight given to each class, which is inversely proportional to the number of samples in each class.     

### Examples

* **Training example using the AF dataset**.  

>   In this example, we will train a heart anomaly detector using the AF dataset [[2]](#af), which contains single-lead ECG recordings and the class labels that categorize the samples into 4 classes: normal rhythm, atrial fibrillation, alternative rhythm and noise. We start by importing the learner and the function used to download and construct the dataset. This dataset has been preprocessed according to [[1]](#gru) and splitted into 5-fold cross validation protocol.  

Reviewer: ```suggestion
In this example, we will train a heart anomaly detector using the AF dataset [[2]](#af), which contains single-lead ECG recordings and the class labels that categorize the samples into 4 classes: normal rhythm, atrial fibrillation, alternative rhythm and noise.
We start by importing the learner and the function used to download and construct the dataset.
This dataset has been preprocessed according to [[1]](#gru) and splitted into 5-fold cross validation protocol.  
```

    If **val_set** is not `None`, it is used to select the model's weights that produce the best validation accuracy.   
  - **test_set**: *engine.datasets.DatasetIterator, default=None*    
    Object that holds the test set.  
    OpenDR dataset object, with `__getitem__` producing a pair of (`engine.data.Timeseries`, `engine.target.Category`).  
  - **class_weight**: *list, default=None*    
    A list that holds the class weight that can be used to counterbalance the effect of imbalanced dataset.   
  - **logging_path**: *str, default=''*    
    Tensorboard path. 
    If not empty, tensorboard data is saved to this path.    
  - **silent**: *bool, default=False*     
    If set to True, disables all printing, otherwise, the performance statistics, estimated time till finish are printed to STDOUT after every epoch.    
  - **verbose**: *bool, default=True*    
    If set to True, enables the progress bar of each epoch.   
 
**Returns**:

  - **performance**: *dict*   
    A dictionary that holds the performance curves with the following keys, with corresponding prefixes (`"train_"`, `"val_"`, `"test_"`):  

>         - `"cross_entropy"`: a list that contains cross entropy measured after each training epoch.     

Reviewer 2: ```suggestion
        - `"cross_entropy"`: a *list* that contains cross entropy measured after each training epoch.     
```

Reviewer 2: Consistency.

        - `"recall"`: a *list* that contains recall measured after each training epoch.   
        - `"f1"`: a *list* that contains f1 measured after each training epoch.     

#### `AttentionNeuralBagOfFeatureLearner.eval`
```python
AttentionNeuralBagOfFeatureLearner.eval(self, dataset, silent, verbose)
```

This method is used to evaluate the current model given the dataset.  
Returns a dictionary containing `"cross_entropy"`, `"acc"`, `"precision"`, `"recall"` and `"f1"` as keys.  
 
**Parameters**:

- **dataset**: *engine.datasets.DatasetIterator*   
  Object that holds the training set.  
  OpenDR dataset object, with `__getitem__` producing a pair of (`engine.data.Timeseries`, `engine.target.Category`).     
- **silent**: *bool, default=False*   
  If set to False, print the cross entropy and accuracy to STDOUT.    
- **verbose**: *bool, default=True*   
  If set to True, display a progress bar of the evaluation process.    
 
**Returns**:

- **performance**: *dict*    
  Dictionary that contains `"cross_entropy"`, `"acc"`, `"precision"`, `"recall"` and `"f1"`.   

#### `AttentionNeuralBagOfFeatureLearner.infer`
```python

> AttentionNeuralBagOfFeatureLearner.infer(series)   

Reviewer 2: ```suggestion
AttentionNeuralBagOfFeatureLearner.infer(self, series)   
```

Reviewer 2: Consistency.

```

This method is used to generate the class prediction given an input series.    
Returns an instance of `engine.target.Category` representing the prediction.    

**Parameters**:

- **series**: *engine.data.Timeseries*   
  Object of type `engine.data.Timeseries` that holds the input data.    
 
**Returns**:

- **prediction**: *engine.target.Category*  
  Object of type `engine.target.Category` that contains the prediction.  

#### `AttentionNeuralBagOfFeatureLearner.save`
```python

> AttentionNeuralBagOfFeatureLearner.save(path, verbose)

Reviewer 2: ```suggestion
AttentionNeuralBagOfFeatureLearner.save(self, path, verbose)
```

```

This method is used to save the current model instance under a given path. 
The saved model can be loaded later by calling `AttentionNeuralBagOfFeatureLearner.load(path)`  
Two files are saved under the given directory path, namely `metadata.json` and `model_weights.pt`. 
The former keeps the metadata and the latter keeps the model weights.

**Parameters**:

- **path**: *str*    
  Directory path to save the model.   
- **verbose**: *bool, default=True*    
  If set to True, print acknowledge message when saving is successful.    

#### `AttentionNeuralBagOfFeatureLearner.load`
```python

> AttentionNeuralBagOfFeatureLearner.load(path, verbose)

Reviewer 2: ```suggestion
AttentionNeuralBagOfFeatureLearner.load(self, path, verbose)
```

```

This method is used to load a previously saved model (by calling `AttentionNeuralBagOfFeatureLearner.save(path)`) from a given directory. 
Note that under the given directory path, `metadata.json` and `model_weights.pt` must exist.   

**Parameters**: 
 
- **path**: *str*   
  Directory path of the model to be loaded.  
- **verbose**: *bool, default=True*     
  If set to True, print acknowledge message when model loading is successful.    

#### `AttentionNeuralBagOfFeatureLearner.download`
```python

> AttentionNeuralBagOfFeatureLearner.download(path, fold_idx)

Reviewer 2: ```suggestion
AttentionNeuralBagOfFeatureLearner.download(self, path, fold_idx)
```

  Whether to standardize the input series.   

**Returns**:  

- **train_set**: *engine.datasets.DatasetIterator*   
  The training set that is constructed from `engine.datasets.DatasetIterator` that is ready to be used with the `fit()` method.    
- **val_set**: *engine.datasets.DatasetIterator*   
  The validation set that is constructed from `engine.datasets.DatasetIterator` that is ready to be used with the `fit()` method.    
- **series_length**: *int*   
  The length of each series (the number of elements in each series).  
  This parameter is used in the constructor of `AttentionNeuralBagOfFeatureLearner`.    
- **class_weight**: *list*   
  The weight given to each class, which is inversely proportional to the number of samples in each class.     

### Examples

* **Training example using the AF dataset**.  
In this example, we will train a heart anomaly detector using the AF dataset [[2]](#af), which contains single-lead ECG recordings and the class labels that categorize the samples into 4 classes: normal rhythm, atrial fibrillation, alternative rhythm and noise.
We start by importing the learner and the function used to download and construct the dataset.

> This dataset has been preprocessed according to [[1]](#gru) and splitted into 5-fold cross validation protocol.  

Reviewer 2: ```suggestion
This dataset has been preprocessed according to [[1]](#anbof) and split into 5-fold cross validation protocol.  
```

  ```python
  learner = AttentionNeuralBagOfFeatureLearner(in_channels=1,
                                               series_length=series_length,
                                               n_class=4,
                                               n_codeword=256,
                                               quantization_type='nbof',
                                               attention_type='temporal',
                                               iters=200,
                                               lr_scheduler=get_cosine_lr_scheduler(2e-4, 1e-5))
  ```

  After the learner has been constructed, we can train the learner using the `fit` function:
   
  ```python
  performance = learner.fit(train_set, val_set, class_weight=class_weight) 
  ```

  In the above code, we pass the `class_weight` parameter to the `fit` function because the AF dataset is imbalanced. The `fit` function returns `performance`, which is a dictionary that contains the performance measured after each training epoch.   

* **Download and evaluate pretrained model for the AF dataset**.  

>   In this example, we will show how pretrained models for the AF dataset [[2]](#af) can be easily downloaded by a single line of code using the functionality provided in `AttentionNeuralBagOfFeatureLearner` and evaluate the model on the test set. Since the AF dataset is divided into 5 folds, the following code iterates through each data split, downloads the corresponding pretrained model and evaluates on the correposnding validation set.  

Reviewer 2: ```suggestion
  In this example, we will show how pretrained models for the AF dataset [[2]](#af) can be easily downloaded by a single line of code using the functionality provided in `AttentionNeuralBagOfFeatureLearner` and evaluate the model on the test set.
  Since the AF dataset is divided into 5 folds, the following code iterates through each data split, downloads the corresponding pretrained model and evaluates on the corresponding validation set.  
```

Reviewer 2: One sentence per line.

  if not os.path.exists(pretrained_model_dir):
      os.mkdir(pretrained_model_dir)

  # sample length of 30 seconds
  sample_length = 30

  # iterate through each validation fold
  # download and evaluate pretrained model
  for fold_idx in range(5):
      train_set, val_set, series_length, class_weight = get_AF_dataset(data_file, fold_idx, sample_length)
      learner = AttentionNeuralBagOfFeatureLearner(in_channels=1,
                                                   series_length=series_length,
                                                   n_class=4,
                                                   n_codeword=512,
                                                   quantization_type='nbof',
                                                   attention_type='temporal')

      # create a sub directory for each pretrained model
      model_path = os.path.join(pretrained_model_dir, 'AF_{}'.format(fold_idx))
      learner.download(model_path, fold_idx)

      # after pretrained model is downloaded to the given path
      # pretrained weights can be loaded with `load()`
      learner.load(model_path)

      # evaluate
      learner.eval(val_set)
  ```

#### References

> <a name="gru" href="https://arxiv.org/abs/2005.12250">[1]</a> Attention-based Neural Bag-of-Feautures Learning for Sequence Data,

Reviewer 2: ```suggestion
<a name="anbof" href="https://arxiv.org/abs/2005.12250">[1]</a> Attention-based Neural Bag-of-Feautures Learning for Sequence Data,
```
